shared/utility/strategy/positioning.hpp

```diff
@@ -0,0 +1,264 @@
+            cost += avoid_weight * penalty(d, clearance);
+        }
+
+        return cost;
+    }
+
+    /**
+     * @brief Find the optimal position for a robot using gradient descent
+     *
+     * @param initial The initial position of the robot
+     * @param neutral_point The neutral point to be close to
+     * @param opponent_positions The positions of the opponent robots
+     * @param all_robot_positions The positions of all robots (including self)
+     * @param clearance The clearance distance to avoid other robots
+     * @param equidist_weight Weight for equidistance cost
+     * @param neutral_weight Weight for closeness to neutral point
+     * @param avoid_weight Weight for avoiding other robots
+     * @param step_size Step size for gradient descent
+     * @param max_iters Maximum number of iterations for gradient descent
+     *
+     * @return The optimal position for the robot
+     */
+    template <typename Scalar>
+    Eigen::Vector3d optimal_pos(const Eigen::Vector3d& initial,
+                                const Eigen::Vector3d& neutral_point,
+                                const std::vector<Eigen::Vector3d>& opponent_positions,
+                                const std::vector<Eigen::Vector3d>& all_robot_positions,
+                                const Scalar clearance,
+                                const Scalar equidist_weight,
+                                const Scalar neutral_weight,
+                                const Scalar avoid_weight,
+                                const Scalar step_size,
+                                const int max_iters) {
+        Eigen::Vector3d X = initial;
+
+        // Gradient descent to find the optimal position
+        for (int iter = 0; iter < max_iters; ++iter) {
+            Eigen::Vector3d grad(0.0, 0.0, 0.0);
+
+            // Numerical gradient
+            Scalar eps = 1e-5;
+            for (int i = 0; i < 2; ++i) {
+                Eigen::Vector3d X_eps = X;
+                X_eps[i] += eps;
+                Scalar f1 = cost(X_eps,
+                                 neutral_point,
+                                 opponent_positions,
+                                 all_robot_positions,
+                                 clearance,
+                                 equidist_weight,
+                                 neutral_weight,
+                                 avoid_weight);
+                X_eps[i] -= 2 * eps;
+                Scalar f2 = cost(X_eps,
+                                 neutral_point,
+                                 opponent_positions,
+                                 all_robot_positions,
+                                 clearance,
+                                 equidist_weight,
+                                 neutral_weight,
+                                 avoid_weight);
+                grad[i]   = (f1 - f2) / (2 * eps);
+            }
+
+            // Update the position
+            X -= step_size * grad;
```

Comment:
I think using the nlopt library would be better suited here rather than implementing this from scratch

It can easily support adding constraints to ensure the optimal position is not outside the field, too far away etc

Also has better algorithms and debugging features out of the box